hazelcast/src/main/java/com/hazelcast/query/impl/CachedQueryEntry.java

```diff
@@ -0,0 +1,195 @@
+import com.hazelcast.internal.serialization.SerializationService;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.nio.serialization.Portable;
+import com.hazelcast.query.QueryException;
+import com.hazelcast.query.impl.getters.ReflectionHelper;
+
+import static com.hazelcast.query.QueryConstants.KEY_ATTRIBUTE_NAME;
+import static com.hazelcast.query.QueryConstants.THIS_ATTRIBUTE_NAME;
+
+/**
+ * Entry of the Query.
+ */
+public class CachedQueryEntry implements QueryableEntry {
+
+    private Data keyData;
+    private Object keyObject;
+    private Data valueData;
+    private Object valueObject;
+    private SerializationService serializationService;
+
+    public CachedQueryEntry() {
+    }
+
+    public CachedQueryEntry(SerializationService serializationService, Data key, Object value) {
+        init(serializationService, key, value);
+    }
+
+    public void init(SerializationService serializationService, Data key, Object value) {
+        if (key == null) {
+            throw new IllegalArgumentException("keyData cannot be null");
+        }
+        this.serializationService = serializationService;
+        this.keyData = key;
+        this.keyObject = null;
+
+        if (value instanceof Data) {
+            this.valueData = (Data) value;
+            this.valueObject = null;
+        } else {
+            this.valueObject = value;
+            this.valueData = null;
+        }
+    }
+
+    @Override
+    public Object getValue() {
+        if (valueObject == null) {
+            valueObject = serializationService.toObject(valueData);
+        }
+        return valueObject;
+    }
+
+    @Override
+    public Object getKey() {
+        if (keyObject == null) {
+            keyObject = serializationService.toObject(keyData);
+        }
+        return keyObject;
+    }
+
+    @Override
+    public Data getKeyData() {
+        if (keyData == null) {
+            keyData = serializationService.toData(keyObject);
+        }
+        return keyData;
+    }
+
+    @Override
+    public Data getValueData() {
+        if (valueData == null) {
+            valueData = serializationService.toData(valueObject);
+        }
+        return valueData;
+    }
+
+
+    @Override
+    public Comparable getAttribute(String attributeName) throws QueryException {
+        if (KEY_ATTRIBUTE_NAME.equals(attributeName)) {
+            return (Comparable) getKey();
+        } else if (THIS_ATTRIBUTE_NAME.equals(attributeName)) {
+            return (Comparable) getValue();
+        }
+
+        boolean key = QueryEntryUtils.isKey(attributeName);
+        attributeName = QueryEntryUtils.getAttributeName(key, attributeName);
+        Object targetObject = getTargetObject(key);
+
+        return QueryEntryUtils.extractAttribute(attributeName, targetObject, serializationService);
+    }
+
+    private Object getTargetObject(boolean key) {
+        Object targetObject;
+        if (key) {
+            if (keyObject == null) {
```

Comment:
Pull out 2 inner methods. getTargetObjectFromKey and getTargetObjectFromValue or something.